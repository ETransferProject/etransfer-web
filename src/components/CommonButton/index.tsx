import { Button, ButtonProps } from 'antd';
import clsx from 'clsx';
import styles from './styles.module.scss';

export enum CommonButtonSize {
  ExtraSmall = 'extraSmall',
  Small = 'small',
  Middle = 'middle',
  Large = 'large',
}

export enum CommonButtonType {
  Primary = 'primary',
  Secondary = 'secondary',
}

export type CommonButtonProps = Omit<ButtonProps, 'size' | 'type'> & {
  size?: CommonButtonSize;
  type?: CommonButtonType;
  stretched?: boolean;
};

export default function CommonButton({
  size = CommonButtonSize.Middle,
  type = CommonButtonType.Primary,
  stretched = false,
  ...props
}: CommonButtonProps) {
  return (
    <Button
      {...props}
      className={clsx(
        styles['common-button'],
        styles[size],
        styles[type],
        { [styles['stretched']]: stretched },
        props.className,
      )}
    />
  );
}
