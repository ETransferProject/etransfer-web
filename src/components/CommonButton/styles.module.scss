@import '/src/assets/theme/color.scss';
@import '/src/assets/theme/font.scss';

.common-button {
  border-radius: 6px;
  box-shadow: none;
  span {
    @extend .font-family-medium;
  }
}

.primary {
  background: $brand;
  border: 1px solid $brand;
  color: $white-text;
  &:hover {
    background: $hover;
    border-color: $hover;
    color: $white-text;
  }
  &:checked {
    background: $click;
    border-color: $click;
    color: $white-text;
  }
  &:disabled,
  &[disabled]:hover {
    background: $disable;
    border-color: $disable;
    color: $white-text;
  }
  &:focus {
    background: $click;
    border-color: $click;
    color: $white-text;
  }
}

.secondary {
  background: $normal-bg;
  border: 1px solid $normal-bg;
  color: $primary-text;
  &:hover {
    background: $hover-bg;
    border-color: $hover-bg;
    color: $primary-text;
  }
  &:checked {
    background: $click-bg;
    border-color: $click-bg;
    color: $primary-text;
  }
  &:disabled {
    background: $hover-bg;
    border-color: $hover-bg;
    color: $primary-text;
  }
  &:focus {
    background: $click-bg;
    border-color: $click-bg;
    color: $primary-text;
  }
}

.border {
  background: $page-bg !important;
  border: 1px solid $border !important;
  color: $primary-text !important;
  display: flex;
  align-items: center;
  &:hover {
    background: $page-bg !important;
    border-color: $hover !important;
    color: $primary-text !important;
  }
  &:checked {
    background: $page-bg;
    border-color: $click;
    color: $primary-text;
  }
  &:disabled {
    background: $disable;
    border-color: $disable;
    color: $primary-text;
  }
  &:focus {
    background: $page-bg;
    border-color: $click;
    color: $primary-text;
  }
}

.extraSmall {
  height: 32px;
  padding: 0 16px;
  span {
    display: inline-flex;
    font-size: 12px;
    line-height: 20px;
  }
}

.small {
  height: 40px;
  padding: 0 20px;
  span {
    display: inline-flex;
    font-size: 14px;
    line-height: 22px;
  }
}

.middle {
  height: 48px;
  padding: 0 28px;
  span {
    display: inline-flex;
    font-size: 16px;
    line-height: 24px;
  }
}

.large {
  height: 56px;
  padding: 0 32px;
  span {
    display: inline-flex;
    font-size: 16px;
    line-height: 24px;
  }
}

.stretched {
  width: 100%;
}
