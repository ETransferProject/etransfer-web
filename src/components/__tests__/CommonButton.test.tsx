import { beforeEach, vi, describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import CommonButton, {
  CommonButtonProps,
  CommonButtonSize,
  CommonButtonType,
} from '../CommonButton';
import clsx from 'clsx';

// Mock Antd Button
vi.mock('antd', async (importOriginal) => {
  // const actual = await importOriginal();
  const actual = await importOriginal<typeof import('antd')>();
  return {
    ...actual,
    Button: vi.fn(({ className, ...props }) => (
      <button {...props} className={className} data-testid="antd-button" />
    )),
  };
});

// Emulate clsx class name generation
vi.mock('clsx', () => ({
  default: vi.fn(() => 'mocked-classname'),
}));

describe('CommonButton Component', () => {
  const defaultProps: CommonButtonProps = {
    size: CommonButtonSize.Middle,
    type: CommonButtonType.Primary,
    stretched: false,
  };

  beforeEach(() => {
    (clsx as any).mockClear();
  });

  test('Correct rendering of default configuration', () => {
    render(<CommonButton>Test Button</CommonButton>);

    // Assert that the button is rendered
    const button = screen.getByTestId('antd-button');

    // Assert that the button text is correct
    expect(button.textContent).toBe('Test Button');
  });

  describe('Class name generation logic', () => {
    const classCases: Array<{
      props: Partial<CommonButtonProps>;
      expectedClasses: string[];
    }> = [
      {
        props: { size: CommonButtonSize.Small },
        expectedClasses: ['etransfer-ui-common-button-Small'],
      },
      {
        props: { type: CommonButtonType.Secondary },
        expectedClasses: ['etransfer-ui-common-button-Secondary'],
      },
      {
        props: { stretched: true },
        expectedClasses: ['etransfer-ui-common-button-stretched'],
      },
      {
        props: { className: 'custom-class' },
        expectedClasses: ['custom-class'],
      },
    ];

    test.each(classCases)('Generates the correct class name when passing $props', ({ props }) => {
      render(
        <CommonButton {...defaultProps} {...props}>
          Button
        </CommonButton>,
      );

      expect(clsx).toHaveBeenCalled();
    });

    test('Merge multiple class names', () => {
      render(
        <CommonButton
          size={CommonButtonSize.Large}
          type={CommonButtonType.Secondary}
          stretched
          className="extra-class">
          Button
        </CommonButton>,
      );

      // TODO random hash
      // Assert that the correct class names are generated
      expect(clsx).toHaveBeenCalledWith(
        '_common-button_42d790',
        '_large_42d790',
        '_secondary_42d790',
        {
          _stretched_42d790: true,
        },
        'extra-class',
      );
      // expect(clsx).toHaveBeenCalledWith(
      //   expect.stringMatching(/^_common-button_[0-9a-fA-F]{6}$/),
      //   // expect.stringMatching(/^_large_[0-9a-fA-F]{6}$/),
      //   expect.stringContaining('_large_'),
      //   expect.stringContaining('large'),
      //   expect.stringContaining('secondary'),
      //   expect.objectContaining({
      //     [expect.stringContaining('_stretched_')]: true,
      //   }),
      //   'extra-class',
      // );
    });
  });

  test('Handling undefined and null values', () => {
    render(
      <CommonButton size={undefined} type={undefined} stretched={undefined}>
        Button
      </CommonButton>,
    );

    // TODO random hash
    // Assert that the correct class names are generated
    expect(clsx).toHaveBeenCalledWith(
      '_common-button_42d790',
      '_middle_42d790',
      '_primary_42d790',
      { ['_stretched_42d790']: false },
      undefined,
    );
  });

  test('Snapshot Matching', () => {
    const { asFragment } = render(
      <CommonButton
        size={CommonButtonSize.Small}
        type={CommonButtonType.Primary}
        stretched
        className="custom-class">
        Submit
      </CommonButton>,
    );

    // Assert that the component matches the snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
