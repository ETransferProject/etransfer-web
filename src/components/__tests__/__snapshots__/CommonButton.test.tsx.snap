// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommonButton Component > Snapshot Matching 1`] = `
<DocumentFragment>
  <button
    class="mocked-classname"
    data-testid="antd-button"
  >
    Submit
  </button>
</DocumentFragment>
`;
